lib/lz4-ruby.rb

```diff
@@ -31,17 +31,42 @@ def self.decompress(input, in_size = nil, encoding = nil)
     return result
   end
 
+  def self.compress_with_dict(input, dict, in_size = nil, encoding = nil)
+    return _compress(input, in_size, false, dict)
+  end
+
+  def self.decompress_with_dict(input, dict, in_size = nil, encoding = nil)
+    in_size = input.bytesize if in_size == nil
+    out_size, varbyte_len = decode_varbyte(input)
+
+    if out_size < 0 || varbyte_len < 0
+      raise "Compressed data is maybe corrupted"
+    end
+
+    result = LZ4Internal::decompress_with_dict(input,
+                                               in_size,
+                                               varbyte_len,
+                                               out_size,
+                                               dict,
+                                               dict.size())
+    result.force_encoding(encoding) if encoding != nil
+
+    return result
+  end
+
   # @deprecated Use {#decompress} and will be removed.
   def self.uncompress(input, in_size = nil)
     return decompress(input, in_size)
   end
 
   private
-  def self._compress(input, in_size, high_compression)
+  def self._compress(input, in_size, high_compression, dict = nil)
```

Comment:
The API to this function is confusing. It implies that `high_compression` and `dict` are orthogonal, but they aren't. Do we need to go through this function. TBH I don't know why it even exists.